### DISCUSSION TITLE
129

### DISCUSSION THREAD

**Discussion by jojoldu** [jojoldu/blog-comments#728] [Comments]
http://jojoldu.tistory.com/129

**Comment by jojoldu** [comment]
작성자: 이상욱

잘보고갑니다

**Comment by jojoldu** [comment]
작성자: 개발자

안녕하세요. 글 잘보고있습니다. 감사합니다.
스프링과의 의존성을 낮추기위해 생성자 Injection을 사용하셨다고 하셨는데 이해가 잘 가지 않아서..
@autowired 등을 사용해서 의존관계를맺으면 Spring Annotation이기때문에 스프링에 의존된다는 의도로 말씀해주신건가요???

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

안녕하세요! 방문해주셔서 감사합니다^^

말씀하신대로 스프링 어노테이션을 사용하고 있기 때문에 스프링 외에 다른 프레임워크에서 위 코드를 전혀 사용하지 못하는 이슈도 있고, 또하나의 이유가 DI (의존성 주입)을 좀 더 편하게 하기 위함입니다.

예를 들어 제가 A Controller에서 A Service를 사용하도록 @Autowired 했다고 가정하겠습니다.
이럴 경우 제가 A Controller의 기능을 테스트 하기 위해 A Service가 아닌 임의의 어떤 테스트 Service를 사용하도록 하고 싶은데 @Autowired를 사용할 경우에 이를 교체하기가 굉장히 어렵습니다. 
물론 스프링프레임워크의 테스트 도구를 사용한다면 가능은 하겠지만,  이렇게 하면 테스트를 할때마다 스프링 컨텍스트를 구동시켜야만 합니다.
생성자 인잭션을 쓰면 new AController(MockService) 로 쉽게 서비스 코드를 교체할 수 있는데 @Autowired는 이런 방식이 어렵습니다.

이런 이유로 사실 IntelliJ에선 노골적으로 코드에서 @Autowired를 사용하면 생성자 인잭션을 쓰라고 권장 메세지를 보내주기도 하고 스프링  문서에서 역시 생성자 인잭션을 권장하고 있습니다.

질문에 대한 답이 되었는지 모르겠습니다^^;
더 궁금한게 있으시면 언제든 편하게 댓글 남겨주세요

**Comment by jojoldu** [comment]
작성자: falhed

포스팅 잘봤습니다. 한가지 궁금한것이 있는데요. /member 아무값도 입력하지 않은경우 error :[]하위로 json이 리턴된다고 적혀있는데요.@Valid어노테이션으로 인해  에러가 발생하면 자동으로 저런형태로 리턴이 되는건가요? 리턴값은 Long형인거같은데...음..궁금하네요

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

안녕하세요!
방문해주셔서 감사합니다.

말씀해주신 경우엔 필수값들이 없기 때문에 @Valid 에서 400 Bad Request를 발생해서 그렇습니다.
Long 타입이 반환되는 경우는 정상적으로 /member 가 처리된 경우에만 가능합니다^^

**Comment by ssw207** [comment]
포스팅 잘봤습니다. 보는도중에 잘 이해가 안되는부분이 있어서 질문드려요!

'휴대폰 번호를 DB 저장시에는 조건 검색이 가능하도록 3개 컬럼에 나눠서 저장한다. (ex: 컬럼1: 010, 컬럼2: 1234, 컬럼3: 5678)'

이렇게 말씀하셨는데 어떤방식으로 조건검색이 진행되서  휴대 전화번호의 컬럼을 나누는건가요??

**Comment by engineer-myoa-2** [comment]
좋은 내용 감사합니다

**Comment by DongHwani** [comment]
매번 좋은 글 감사합니다.
하나만 질문드립니다.
 RestController에서 Errors 선언없이 어떻게 error객체를 Long타입으로 반환하여 ajax 통신하는지 궁금합니다.

**Comment by Y-k-Y** [comment]
언제나 좋은 글 감사합니다🙇🏻
